lib/src/markdown_processor.dart

```diff
@@ -560,6 +467,122 @@ ModelElement _findRefElementInLibrary(String codeRef, Warnable element,
+      }
+    }
+  }
+  results.remove(null);
+}
+
+void _findPartiallyQualifiedMatches(Set<ModelElement> results, Library library, String codeRef, String codeRefChomped, PackageGraph packageGraph, Class preferredClass) {
+  // Only look for partially qualified matches if we didn't find a fully qualified one.
+  if (results.isEmpty && library.modelElementsNameMap.containsKey(codeRefChomped)) {
+    for (final modelElement in library.modelElementsNameMap[codeRefChomped]) {
+      if (!_ConsiderIfConstructor(codeRef, modelElement)) continue;
+      results.add(packageGraph.findCanonicalModelElementFor(
+          modelElement.element,
+          preferredClass: preferredClass));
+    }
+  }
+  results.remove(null);
+}
+
+void _findWithinRefElementCache(Set<ModelElement> results, String codeRefChomped, PackageGraph packageGraph, String codeRef) {
+  // We now need the ref element cache to keep from repeatedly searching [Package.allModelElements].
+  // But if not, look for a fully qualified match.  (That only makes sense
+  // if the codeRef might be qualified, and contains periods.)
+  if (results.isEmpty &&
+      codeRefChomped.contains('.') &&
+      packageGraph.findRefElementCache.containsKey(codeRefChomped)) {
+    for (final ModelElement modelElement
+        in packageGraph.findRefElementCache[codeRefChomped]) {
+      if (!_ConsiderIfConstructor(codeRef, modelElement)) continue;
+      // For fully qualified matches, the original preferredClass passed
+      // might make no sense.  Instead, use the enclosing class from the
+      // element in [_findRefElementCache], because that element's enclosing
+      // class will be preferred from [codeRefChomped]'s perspective.
+      results.add(packageGraph.findCanonicalModelElementFor(
+          modelElement.element,
+          preferredClass: modelElement.enclosingElement is Class
+              ? modelElement.enclosingElement
+              : null));
+    }
+  }
+  results.remove(null);
+}
+
+void _findWithinTryClasses(Set<ModelElement> results, Class preferredClass, Warnable element, String codeRefChomped, String codeRef, PackageGraph packageGraph) {
+  if (results.isEmpty) {
+    // Maybe this is local to a class.
+    // TODO(jcollins-g): tryClasses is a strict subset of the superclass chain.  Optimize.
+    List<Class> tryClasses = [preferredClass];
+    Class realClass = tryClasses.first;
+    if (element is Inheritable) {
+      Inheritable overriddenElement = element.overriddenElement;
+      while (overriddenElement != null) {
+        tryClasses.add(
+            (element.overriddenElement as EnclosedElement).enclosingElement);
+        overriddenElement = overriddenElement.overriddenElement;
+      }
+    }
+
+    for (Class tryClass in tryClasses) {
+      if (tryClass != null) {
+        _getResultsForClass(
+            tryClass, codeRefChomped, results, codeRef, packageGraph);
+      }
+      results.remove(null);
```

Comment:
What does removing `null` here do?

Reply:
Nothing all that useful most of the time.  Some of the blocks are capable of adding nulls to the results set, but instead of trying to determine exactly which ones could possibly add a null sometimes we just stamp it out every time.

The refactor left these remove(null) calls in some of the functions.  Reorganized this dance to be a little more consistent and added a TODO to get rid of it.